From: Allan Ievers <aimail-dspam_users@rearden.com>
Subject: Fix recipient corruption when releasing a message from quarantine

When releasing mail from quarantine, dspam corrupts the FROM part in the
SMTP/LMTP handshake.

Origin: vendor
Bug: https://sourceforge.net/p/dspam/bug-tracker/160/
Bug-Debian: http://bugs.debian.org/698136
Last-Update: 2013-09-10

--- a/src/dspam.c
+++ b/src/dspam.c
@@ -499,8 +499,9 @@ process_message (
                 ATX->train_pristine = 1;
         }
 
-        /* Change also the mail recipient */
-        ATX->recipient = CTX->username;
+        /* Change also the mail recipient. ATX->recipient either points to
+	 * recipient[] or mailbox[] in process_users, hence the size of 256 */
+        strlcpy(ATX->recipient, CTX->username, 256);
 
       }
     }
@@ -1634,6 +1635,7 @@ int process_users(AGENT_CTX *ATX, buffer *message) {
     char filename[MAX_FILENAME_LENGTH];
     int optin, optout;
     char *username = NULL;
+    char recipient[256];
 
     /* If ServerParameters specifies a --user, there will only be one
      * instance on the stack, but possible multiple recipients. So we
@@ -1672,7 +1674,7 @@ int process_users(AGENT_CTX *ATX, buffer *message) {
 	username = node_nt->ptr;
 
     if (node_rcpt) {
-      ATX->recipient = node_rcpt->ptr;
+      strlcpy(recipient, node_rcpt->ptr, sizeof(recipient));
       node_rcpt = c_nt_next (ATX->recipients, &c_rcpt);
     } else {
 
@@ -1680,8 +1682,9 @@ int process_users(AGENT_CTX *ATX, buffer *message) {
       if (have_rcpts)
         break;
 
-      ATX->recipient = node_nt->ptr;
+      strlcpy(recipient, node_nt->ptr, sizeof(recipient));
     }
+    ATX->recipient = recipient;
 
       /* If support for "+detail" is enabled, save full mailbox name for
          delivery and strip detail for processing */
